### ISSUE TITLE
Get timeout on querying version from Kafka Server

### ISSUE THREAD

**Issue by slashmili** [kafka4beam/kafka_protocol#89]
I have a problem connecting to a Kafka 2.5.0 Server.

Just to set the stage right, I'm trying to connect to a server that has OAUTHBEARER as the SASL mechanism. We are trying to use a custom handler to support this auth model but the client fails before reaching that during querying version

I'm trying to connect to Mercedes-Benz's Kafka(https://developer.mercedes-benz.com/products/connect_your_fleet/docs)

I'll share as much of public information that I can:

1. Host:PORT: bootstrap.msg-stream.connect-business.net:443
2. Certfile: is downloadable from Firefox by visiting https://bootstrap.msg-stream.connect-business.net

![download cert 01](https://user-images.githubusercontent.com/585764/111609855-036d1480-87db-11eb-9d40-4ba904d397e3.png)

![download cert 02](https://user-images.githubusercontent.com/585764/111609873-07009b80-87db-11eb-91d0-5700d2df4c1d.png)

```elixir
# Erlang version:
1> Bootstrap = [{<<"bootstrap.msg-stream.connect-business.net">>, 443}].
2> Client = my_kafka_client.
3> Opts = [{query_api_versions, true}, {ssl, [{cacertfile, <<"kafka-cluster-external-prod-kafka.pem">>}]}].
4> brod:start_client(Bootstrap, Client, Opts).


# Elixir version:
iex(1)> (1bootstrap = [{"bootstrap.msg-stream.connect-business.net", 443}]
iex(2)> opts = [{:query_api_versions, true}, {:ssl, [{:cacertfile, "kafka-cluster-external-prod-kafka.pem"}]}]
iex(3)> :brod.start_client(bootstrap, :my_kafka_client, opts)
:ok

10:39:53.233 [info]      :supervisor: {:local, :brod_sup}
    :started: [
  pid: #PID<0.208.0>,
  id: :my_kafka_client,
  mfargs: {:brod_client, :start_link,
   [
     [{"bootstrap.msg-stream.connect-business.net", 443}],
     :my_kafka_client,
     [query_api_versions: true, ssl: [cacertfile: "kafka-cluster-external-prod-kafka.pem"]]
   ]},
  restart_type: {:permanent, 10},
  shutdown: 5000,
  child_type: :worker
]
iex(4)> CorrId : 1716395828
ReqIoData : [<<0,18>>,
             <<0,0>>,
             <<102,78,31,52>>,
             [<<0,15>>,<<"my_kafka_client">>],
             []]

10:39:58.589 [warn]  :brod_client [#PID<0.208.0>] :my_kafka_client is terminating
reason: [
  {{"bootstrap.msg-stream.connect-business.net", 443},
   {{{:kpro_req, #Reference<0.602024089.3602382851.100616>, :api_versions, 0,
      false, []}, :timeout},
    [
      {:kpro_lib, :send_and_recv_raw, 4, [file: 'src/kpro_lib.erl', line: 70]},
      {:kpro_lib, :send_and_recv, 5, [file: 'src/kpro_lib.erl', line: 83]},
      {:kpro_connection, :query_api_versions, 4,
       [file: 'src/kpro_connection.erl', line: 246]},
      {:kpro_connection, :init_connection, 2,
       [file: 'src/kpro_connection.erl', line: 233]},
      {:kpro_connection, :init, 4, [file: 'src/kpro_connection.erl', line: 170]},
      {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 226]}
    ]}}
]
```

FYI I added few io:format in `src/kpro_lib.erl` to see the data that goes through the wire
```
-spec send_and_recv(kpro:req(), port(), module(),
                    kpro:client_id(), timeout()) -> kpro:struct().
send_and_recv(#kpro_req{api = API, vsn = Vsn} = Req,
                 Sock, Mod, ClientId, Timeout) ->
  CorrId = make_corr_id(),
  io:format("CorrId : ~p~n",[CorrId]),
  ReqIoData = kpro_req_lib:encode(ClientId, CorrId, Req),
  io:format("ReqIoData : ~p~n",[ReqIoData]),
  try
    RspBin = send_and_recv_raw(ReqIoData, Sock, Mod, Timeout),
    io:format("RspBin : ~p~n",[RspBin]),
    {CorrId, Body} = decode_corr_id(RspBin), %% assert match CorrId
    io:format("Body : ~p~n",[Body]),
    #kpro_rsp{api = API, vsn = Vsn, msg = Msg} = %% assert match API and Vsn

```

At the end I should mentioned that I'm able to connect to this server using Kafka-python library.

**Comment by zmstone** [comment]
Hi @slashmili 
since you are trying things out with modified code.
could you try to see if sasl auth can complete?
if you set `query_api_versions` option to false, it will skip API query and proceed to sasl auth.

if it works fine, try to move the api version query code to *after* sasl auth.

**Comment by slashmili** [comment]
so if I understood you correctly, I disabled query_api_versions and tried using `scram_sha_256`. 

First I tired this setting using [erlkaf](https://github.com/silviucpp/erlkaf). I got this error :
```
07:55:30.757 [error] rdkafka#consumer-1 ERROR [thrd:app]: rdkafka#consumer-1: sasl_ssl://bootstrap.msg-stream.connect-business.net:443/bootstrap: SASL SCRAM-SHA-256 mechanism handshake failed: Local: Broker transport failure: broker's supported mechanisms: (n/a) (after 0ms in state DOWN)
07:55:35.823 [error] rdkafka#consumer-1 FAIL [thrd:sasl_ssl://bootstrap.msg-stream.connect-business.net:443/bootst]: sasl_ssl://bootstrap.msg-stream.connect-business.net:443/bootstrap: SASL SCRAM-SHA-256 mechanism handshake failed: Broker: Unsupported SASL mechanism: broker's supported mechanisms: OAUTHBEARER (after 320ms in state AUTH_HANDSHAKE)
```
librdkafka is offering  SCRAM-SHA-256 SASL AUTH to the server. After initial handshake it gets a message from the server that it only supports OAUTHBEARER

Then I tried using kafka_protocol but it resulted in timeout.

```elixir
iex(1)> bootstrap = [{"bootstrap.msg-stream.connect-business.net", 443}]
iex(2)> opts = [{:query_api_versions, false}, {:sasl, {:scram_sha_256, "user", "pass"}}, {:ssl, [{:cacertfile, "kafka-cluster-external-prod-kafka.pem"}]}]
iex(3)> :brod.start_client(bootstrap, :my_kafka_client, opts)


09:49:50.872 [info]      :supervisor: {:local, :brod_sup}
    :started: [
  pid: #PID<0.232.0>,
  id: :my_kafka_client,
  mfargs: {:brod_client, :start_link,
   [
     [{"bootstrap.msg-stream.connect-business.net", 443}],
     :my_kafka_client,
     [
       sasl: {:scram_sha_256, "user",
        #Function<11.111811661/0 in :brod_utils.init_sasl_opt/1>},
       query_api_versions: false,
       ssl: [cacertfile: "kafka-cluster-external-prod-kafka.pem"]
     ]
   ]},
  restart_type: {:permanent, 10},
  shutdown: 5000,
  child_type: :worker
]
:ok
 CorrId : 1921034877
ReqIoData : [<<0,17>>,
             <<0,0>>,
             <<114,128,170,125>>,
             [<<0,15>>,<<"my_kafka_client">>],
             [[<<0,13>>,<<"SCRAM-SHA-256">>]]]

 
09:49:56.235 [warn]  :brod_client [#PID<0.232.0>] :my_kafka_client is terminating
reason: [
  {{"bootstrap.msg-stream.connect-business.net", 443},
   {{{:kpro_req, #Reference<0.3111728915.1938554891.133845>, :sasl_handshake, 0,
      false, [[<<0, 13>>, "SCRAM-SHA-256"]]}, :timeout},
    [
      {:kpro_lib, :send_and_recv_raw, 4, [file: 'src/kpro_lib.erl', line: 70]},
      {:kpro_lib, :send_and_recv, 5, [file: 'src/kpro_lib.erl', line: 83]},
      {:kpro_sasl, :handshake, 6, [file: 'src/kpro_sasl.erl', line: 117]},
      {:kpro_sasl, :auth, 7, [file: 'src/kpro_sasl.erl', line: 47]},
      {:kpro_connection, :init_connection, 2,
       [file: 'src/kpro_connection.erl', line: 240]},
      {:kpro_connection, :init, 4, [file: 'src/kpro_connection.erl', line: 170]},
      {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 226]}
    ]}}
]

```

**Comment by zmstone** [comment]
I'm afraid this is a rabbit hole which will take time to investigate.
I have not been following up on Kafka's auth API updates

By any chance you can get (or reproduce) the broker side logs ?

**Comment by slashmili** [comment]
@zmstone I understand if it's hard for you to debug it without having server logs.

However I wouldn't say this is an auth API issue, all the clients I tried are able to query the version from this server(by using the same pen file) except kafka_protocol library so something more fundamental is wrong, something in the protocol level or maybe ssl level 🤷

**Comment by zmstone** [comment]
I'm not saying it's an auth API issue. And I am not blaming server for it.

Anyway, when I test it, the socket is closed (not timeout). with latest `master` branch of `brod`.
brod commit: 42cd15d
Erlnag/OTP version: 23.2.2
```
> make brod-cli
> ./_build/brod_cli/bin/brod_cli meta -b bootstrap.msg-stream.connect-business.net:443 --ssl --cacertfile /tmp/ca.pem
*** throw:{failed_to_connect,
              [{{"bootstrap.msg-stream.connect-business.net",443},
                {{{kpro_req,#Ref<0.1466540807.2868379652.235415>,
                      api_versions,0,false,[]},
                  closed},
                 [{kpro_lib,send_and_recv_raw,4,
                      [{file,"kpro_lib.erl"},{line,70}]},
                  {kpro_lib,send_and_recv,5,[{file,"kpro_lib.erl"},{line,81}]},
                  {kpro_connection,query_api_versions,4,
                      [{file,"kpro_connection.erl"},{line,246}]},
                  {kpro_connection,init_connection,2,
                      [{file,"kpro_connection.erl"},{line,233}]},
                  {kpro_connection,init,4,
                      [{file,"kpro_connection.erl"},{line,170}]},
                  {proc_lib,init_p_do_apply,3,
                      [{file,"proc_lib.erl"},{line,226}]}]}}]}
[{kpro_brokers,with_connection,3,[{file,"kpro_brokers.erl"},{line,62}]},
 {brod_cli,run,5,[{file,"brod_cli.erl"},{line,417}]},
 {brod_cli,main,5,[{file,"brod_cli.erl"},{line,392}]},
 {escript,run,2,[{file,"escript.erl"},{line,758}]},
 {escript,start,1,[{file,"escript.erl"},{line,277}]},
 {init,start_em,1,[]},
 {init,do_boot,3,[]}]
```

And it's the same with or without the `--cacertfile` option

I'll maybe try another API query version when I get time for it.

**Comment by zmstone** [comment]
Timeout was due to `kafka_protocol` is expecting it to be a `{packet, 4}` protocol
but it's in fact HTTP, these are the raw bytes returned from `{"bootstrap.msg-stream.connect-business.net", 443}`

```
<<"HTTP/1.1 400 Bad Request\r\nDate: Sun, 21 Mar 2021 13:41:00 GMT\r\nContent-Type: text/html\r\nContent-Length: 150\r\nConnection: close\r\n\r\n<html>\r\n<head><title>400 Bad Request</title></head>\r\n<body>\r\n<center><h1>400 Bad Request</h1></center>\r\n<hr><center>nginx</center>\r\n</body>\r\n</html>\r\n">>
```
Please connect a Kafka listener. not HTTP listener.

**Comment by slashmili** [comment]
@zmstone thanks for spending time and looking into the issue we have.

Unfortunately I don't see the same behaviour as you experienced. I still see `timeout` using [1c9144ba5a962@brod](https://github.com/kafka4beam/brod/commit/1c9144ba5a96267fe074f9ff04e3f76490bfe6bc).

The only way I can explain it that there is a proxy server in front of this Kafka Server and during your testing something went wrong with the server!

```
./_build/brod_cli/bin/brod_cli meta -b bootstrap.msg-stream.connect-business.net:443 --ssl
*** throw:{failed_to_connect,
              [{{"bootstrap.msg-stream.connect-business.net",443},
                {{{kpro_req,#Ref<0.3949518473.3446407175.20537>,api_versions,
                      0,false,[]},
                  timeout},
                 [{kpro_lib,send_and_recv_raw,4,
                      [{file,
                           "/private/tmp/brod/_build/default/lib/kafka_protocol/src/kpro_lib.erl"},
                       {line,70}]},
                  {kpro_lib,send_and_recv,5,
                      [{file,
                           "/private/tmp/brod/_build/default/lib/kafka_protocol/src/kpro_lib.erl"},
                       {line,81}]},
                  {kpro_connection,query_api_versions,4,
                      [{file,
                           "/private/tmp/brod/_build/default/lib/kafka_protocol/src/kpro_connection.erl"},
                       {line,246}]},
                  {kpro_connection,init_connection,2,
                      [{file,
                           "/private/tmp/brod/_build/default/lib/kafka_protocol/src/kpro_connection.erl"},
                       {line,233}]},
                  {kpro_connection,init,4,
                      [{file,
                           "/private/tmp/brod/_build/default/lib/kafka_protocol/src/kpro_connection.erl"},
                       {line,170}]},
                  {proc_lib,init_p_do_apply,3,
                      [{file,"proc_lib.erl"},{line,226}]}]}}]}
[{kpro_brokers,with_connection,3,
               [{file,"/private/tmp/brod/_build/default/lib/kafka_protocol/src/kpro_brokers.erl"},
                {line,62}]},
 {brod_cli,run,5,[{file,"/private/tmp/brod/src/brod_cli.erl"},{line,417}]},
 {brod_cli,main,5,[{file,"/private/tmp/brod/src/brod_cli.erl"},{line,392}]},
 {escript,run,2,[{file,"escript.erl"},{line,758}]},
 {escript,start,1,[{file,"escript.erl"},{line,277}]},
 {init,start_em,1,[]},
 {init,do_boot,3,[]}]
```

```
./_build/brod_cli/bin/brod_cli meta -b bootstrap.msg-stream.connect-business.net:443 --ssl --cacertfile ~/Downloads/kafka-cluster-external-prod-kafka.pem
*** throw:{failed_to_connect,
              [{{"bootstrap.msg-stream.connect-business.net",443},
                {{{kpro_req,#Ref<0.714483818.1835794445.81148>,api_versions,
                      0,false,[]},
                  timeout},
                 [{kpro_lib,send_and_recv_raw,4,
                      [{file,
```

```
 curl -v -k https://bootstrap.msg-stream.connect-business.net
*   Trying 51.138.33.2...
* TCP_NODELAY set
* Connected to bootstrap.msg-stream.connect-business.net (51.138.33.2) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: O=io.strimzi; CN=kafka-cluster-external-prod-kafka
*  start date: Sep 23 14:17:37 2020 GMT
*  expire date: Sep 22 14:17:37 2025 GMT
*  issuer: O=io.strimzi; CN=cluster-ca v0
*  SSL certificate verify result: self signed certificate in certificate chain (19), continuing anyway.
> GET / HTTP/1.1
> Host: bootstrap.msg-stream.connect-business.net
> User-Agent: curl/7.64.1
> Accept: */*
>
* TLSv1.2 (IN), TLS alert, close notify (256):
* Empty reply from server
* Connection #0 to host boo

```

If you don't mind I reopen this issue and will try to setup a local server that we can reproduce this issue in a more controlled environment.

UPDATE: I can't reopen the issue! Would you please do it?

**Comment by zmstone** [comment]
I inspected the raw bytes returned from the server, and it's HTTP 400.
Kafka is clearly not accepting API version query right after TLS handshakes.

One explanation could be the bytes were returned from the oauth backend which get 'proxied' back the `kafka_protocol`
But it's really hard to reason why should this change the fundamental `{packet, 4}` protocol.

If you want to play with it, I have pushed my test code here: https://github.com/zmstone/kafka_protocol/commit/5b977787ac2cb1b58781403714ea9be95af2dd1a
